# Importing essential libraries

In [28]:
import selenium
import time
from selenium import webdriver

from bs4 import BeautifulSoup
import pandas as pd

# Loading the `Webdriver`

In [47]:
driver = webdriver.Chrome()

# Getting the access of `ACT Gasfitter's` official website

In [48]:
driver.get('https://www.accesscanberra.act.gov.au/s/public-registers/construction-licences?registerid=m0I4a00000000CLEAY')
time.sleep(2)

# Finding number of pages of the data 

In [49]:
pages=driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/p').text
pages

'Page: 1 / 167'

In [50]:
numbers = []
for word in pages.split():
    if word.isdigit():
        numbers.append(int(word))
print(numbers[1])

167


# Main program to get `Headers` & `all the data rows` of just the `first page`

In [51]:
page_source = driver.page_source
soup =  BeautifulSoup(page_source, 'lxml')
            
data = [j.text for j in soup.findAll('td')]
#print(data)
all_names= []
all_names= data[::5]
print(all_names)

['A & A HOTWATER & GAS SOLUTIONS PTY LTD', 'A & A HOTWATER & GAS SOLUTIONS PTY LTD', 'A & K (ACT) PTY LTD', 'A PLUS PLUMBING (ACT) PTY LTD', 'ABBEY, DAMIEN LETTON', 'ABBEY, DAMIEN LETTON', 'ABDEL RAHMAN, MUDATHIR NOURIN ALI', 'ABSOLOM, ALEX', 'ACKLAND, CHRISTOPHER LEE', 'ADAMS, JAMES']


In [52]:
len(all_names)

10

In [53]:
data= data[3:]
#data
all_class_desc= []
all_class_desc= data[::5]
all_class_desc

['LPG Gasfitter (Vapour Phase)',
 'General',
 'General',
 'Advanced',
 'General',
 'LPG Gasfitter (Vapour Phase)',
 'Journeyperson Gasfitter',
 'Journeyperson Gasfitter',
 'General',
 'Journeyperson Gasfitter']

# Generating xpaths for all the records in any page

In [54]:
xpaths=[]
for j in range(1,11): #creating xpaths: just change span[i] to click different links
    xpaths.append('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr['+str(j) + ']/td[5]/lightning-button/button')
print(xpaths) 

['/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[1]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[2]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[3]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[4]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[5]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/di

In [55]:
licensee_detail=[]
for k in range(1,11): #getting the details of every license in a combination
    driver.find_element_by_xpath(xpaths[k-1]).click()
    #time.sleep(1)
    
    page_source = driver.page_source
    soup = BeautifulSoup(page_source, 'lxml')
            
    det= [j.text for j in soup.findAll('td')]
    det=det[1:]
    #det

    ab=[]
    ab=det[::2]

    ab.insert(0,all_names[k-1])
    ab.insert(3,all_class_desc[k-1])
    ab

    licensee_detail.append(ab)
    
    
    
    
    
    driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div/lightning-button/button').click()

In [56]:
licensee_detail[0]

['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
 '20191321',
 '2/2/2024',
 'LPG Gasfitter (Vapour Phase)',
 'Gasfitter',
 '',
 '']

In [57]:
len(licensee_detail)

10

# Main program to get `Headers` & `all the data rows` of all the pages till 167

In [58]:
for i in range(1, numbers[1]):
    
    driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/div[4]/button').click()
    time.sleep(1)
    
    page_source = driver.page_source
    soup =  BeautifulSoup(page_source, 'lxml')       
    data = [j.text for j in soup.findAll('td')]
    #data
    all_names= []
    all_names= data[::5]
    #all_names
    data= data[3:]
    #data
    all_class_desc= []
    all_class_desc= data[::5]
    #all_class_desc
    
    
    for k in range(1,11): 
        
        driver.find_element_by_xpath(xpaths[k-1]).click()
        #time.sleep(1)
    
        page_source = driver.page_source
        soup = BeautifulSoup(page_source, 'lxml')
            
        det= [l.text for l in soup.findAll('td')]
        det=det[1:]
        #det

        ab=[]
        ab=det[::2]

        ab.insert(3,all_class_desc[k-1])
        ab.insert(0,all_names[k-1])
        #ab

        licensee_detail.append(ab)
    
        driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div/lightning-button/button').click()

NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[10]/td[5]/lightning-button/button"}
  (Session info: chrome=94.0.4606.71)


In [59]:
det

['201583',
 'Expiry date',
 '1/6/2024',
 'Occupation',
 'Gasfitter',
 'General',
 '',
 'LPG Gasfitter (Vapour Phase)',
 '']

In [60]:
ab

['ZOUCH, GEORGE ANDREW',
 '201583',
 '1/6/2024',
 'Gasfitter',
 'LPG Gasfitter (Vapour Phase)',
 '',
 '']

In [61]:
licensee_detail

[['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
  '20191321',
  '2/2/2024',
  'LPG Gasfitter (Vapour Phase)',
  'Gasfitter',
  '',
  ''],
 ['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
  '20191321',
  '2/2/2024',
  'General',
  'Gasfitter',
  '',
  ''],
 ['A & K (ACT) PTY LTD', '2016855', '13/9/2022', 'General', 'Gasfitter', ''],
 ['A PLUS PLUMBING (ACT) PTY LTD',
  '2016683',
  '17/8/2022',
  'Advanced',
  'Gasfitter',
  ''],
 ['ABBEY, DAMIEN LETTON',
  '200216119',
  '17/2/2024',
  'General',
  'Gasfitter',
  '',
  ''],
 ['ABBEY, DAMIEN LETTON',
  '200216119',
  '17/2/2024',
  'LPG Gasfitter (Vapour Phase)',
  'Gasfitter',
  '',
  ''],
 ['ABDEL RAHMAN, MUDATHIR NOURIN ALI',
  '20201405',
  '31/10/2023',
  'Journeyperson Gasfitter',
  'Gasfitter',
  ''],
 ['ABSOLOM, ALEX',
  '2021915',
  '27/7/2024',
  'Journeyperson Gasfitter',
  'Gasfitter',
  ''],
 ['ACKLAND, CHRISTOPHER LEE',
  '2014830',
  '5/4/2023',
  'General',
  'Gasfitter',
  ''],
 ['ADAMS, JAMES',
  '2021668',
  '8/8/2024',
  'Jou

In [62]:
licensee_detail[0]

['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
 '20191321',
 '2/2/2024',
 'LPG Gasfitter (Vapour Phase)',
 'Gasfitter',
 '',
 '']

In [63]:
len(licensee_detail)

1669

# Creating the final dataframe by zipping all the attribute's lists

In [64]:
df = pd.DataFrame(licensee_detail, columns = ['Name', 'License number','Expiry Date','Class description','Occupation','x1','x2'])
 
# print dataframe.
df

Name License number  \
0            A & A HOTWATER & GAS SOLUTIONS PTY LTD       20191321   
1            A & A HOTWATER & GAS SOLUTIONS PTY LTD       20191321   
2                               A & K (ACT) PTY LTD        2016855   
3                     A PLUS PLUMBING (ACT) PTY LTD        2016683   
4                              ABBEY, DAMIEN LETTON      200216119   
5                              ABBEY, DAMIEN LETTON      200216119   
6                 ABDEL RAHMAN, MUDATHIR NOURIN ALI       20201405   
7                                     ABSOLOM, ALEX        2021915   
8                          ACKLAND, CHRISTOPHER LEE        2014830   
9                                      ADAMS, JAMES        2021668   
10    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD        2017904   
11                                 AGGARWAL, SURESH          20151   
12                           AGOSTINO, ALEX DOMINIC       20121662   
13                                  AL-KULABI, SAIF        2020367   
14                                  AL-KULABI, SAIF        2020367   
15                                AL-ZAHROUT, BRAKH       20211300   
16                 ALL GAS PLUMBING REPAIRS PTY LTD          20208   
17                 ALL GAS PLUMBING REPAIRS PTY LTD          20208   
18                             ALLEN, GAVAN CHARLES       20151235   
19                             ALLEN, GERARD THOMAS       20101061   
20                            ALLEN, MATHEW CHARLES      200428226   
21                              AMEY, BRETT RICHARD      199915913   
22                       AMEY, SAMUEL LESLIE THOMAS      200216054   
23                              ANDERSON, DEAN JOHN      199214659   
24                            ANDREW, SCOTT MATTHEW        2010188   
25                            ANDREW, SCOTT MATTHEW        2010188   
26                             ANDREWS, JUSTIN PAUL      200416233   
27                             ANDREWS, JUSTIN PAUL      200416233   
28                                 ANTONAKOS, CHRIS      199714524   
29                                     APAP, DANIEL        2021843   
...                                             ...            ...   
1639                    WOODBRIDGE, MATTHEW STEPHEN      199815827   
1640                         WOODBURY, MALCOLM JOHN      199714937   
1641                            WOODFORD, RYAN SEAN        2016267   
1642                                WOODS, BENJAMIN       20071006   
1643                                WOODS, BENJAMIN       20071006   
1644                                  WOODS, ELLIOT       20201163   
1645                        WOODWARD, ADAM KINGSLEY      199714141   
1646                        WOODWARD, ADAM KINGSLEY      199714141   
1647                          WOODWARD, MARK VICTOR      200316156   
1648                                    WRIGHT, BEN        2021343   
1649                          WRIGHT, JAMES BERNARD        2009649   
1650                          WRIGHT, JAMES BERNARD        2009649   
1651                                  WRIGHT, OSCAR        2020897   
1652                         WRIGHT, WILLIAM JOSEPH        2014616   
1653                         WRIGHT, WILLIAM JOSEPH        2014616   
1654                        WROE, CARRINGTON ROBERT        2007277   
1655                        WROE, CARRINGTON ROBERT        2007277   
1656                                  WYBORN, ETHAN       20181262   
1657                                  WYBORN, ETHAN       20181262   
1658                                   WYBURN, SEAN        2017261   
1659                            XL PLUMBING PTY LTD        2008404   
1660                            XL PLUMBING PTY LTD        2008404   
1661              YOUR LOCAL PLUMBING GROUP PTY LTD        2016603   
1662                                 ZAMMIT, EDWARD      199613986   
1663                                   ZHANG, CHENG        2021181   
1664                            ZINGLEMANN, GREGORY      

In [65]:
df = df.drop(['x1','x2'], 1)
df

Name License number  \
0            A & A HOTWATER & GAS SOLUTIONS PTY LTD       20191321   
1            A & A HOTWATER & GAS SOLUTIONS PTY LTD       20191321   
2                               A & K (ACT) PTY LTD        2016855   
3                     A PLUS PLUMBING (ACT) PTY LTD        2016683   
4                              ABBEY, DAMIEN LETTON      200216119   
5                              ABBEY, DAMIEN LETTON      200216119   
6                 ABDEL RAHMAN, MUDATHIR NOURIN ALI       20201405   
7                                     ABSOLOM, ALEX        2021915   
8                          ACKLAND, CHRISTOPHER LEE        2014830   
9                                      ADAMS, JAMES        2021668   
10    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD        2017904   
11                                 AGGARWAL, SURESH          20151   
12                           AGOSTINO, ALEX DOMINIC       20121662   
13                                  AL-KULABI, SAIF        2020367   
14                                  AL-KULABI, SAIF        2020367   
15                                AL-ZAHROUT, BRAKH       20211300   
16                 ALL GAS PLUMBING REPAIRS PTY LTD          20208   
17                 ALL GAS PLUMBING REPAIRS PTY LTD          20208   
18                             ALLEN, GAVAN CHARLES       20151235   
19                             ALLEN, GERARD THOMAS       20101061   
20                            ALLEN, MATHEW CHARLES      200428226   
21                              AMEY, BRETT RICHARD      199915913   
22                       AMEY, SAMUEL LESLIE THOMAS      200216054   
23                              ANDERSON, DEAN JOHN      199214659   
24                            ANDREW, SCOTT MATTHEW        2010188   
25                            ANDREW, SCOTT MATTHEW        2010188   
26                             ANDREWS, JUSTIN PAUL      200416233   
27                             ANDREWS, JUSTIN PAUL      200416233   
28                                 ANTONAKOS, CHRIS      199714524   
29                                     APAP, DANIEL        2021843   
...                                             ...            ...   
1639                    WOODBRIDGE, MATTHEW STEPHEN      199815827   
1640                         WOODBURY, MALCOLM JOHN      199714937   
1641                            WOODFORD, RYAN SEAN        2016267   
1642                                WOODS, BENJAMIN       20071006   
1643                                WOODS, BENJAMIN       20071006   
1644                                  WOODS, ELLIOT       20201163   
1645                        WOODWARD, ADAM KINGSLEY      199714141   
1646                        WOODWARD, ADAM KINGSLEY      199714141   
1647                          WOODWARD, MARK VICTOR      200316156   
1648                                    WRIGHT, BEN        2021343   
1649                          WRIGHT, JAMES BERNARD        2009649   
1650                          WRIGHT, JAMES BERNARD        2009649   
1651                                  WRIGHT, OSCAR        2020897   
1652                         WRIGHT, WILLIAM JOSEPH        2014616   
1653                         WRIGHT, WILLIAM JOSEPH        2014616   
1654                        WROE, CARRINGTON ROBERT        2007277   
1655                        WROE, CARRINGTON ROBERT        2007277   
1656                                  WYBORN, ETHAN       20181262   
1657                                  WYBORN, ETHAN       20181262   
1658                                   WYBURN, SEAN        2017261   
1659                            XL PLUMBING PTY LTD        2008404   
1660                            XL PLUMBING PTY LTD        2008404   
1661              YOUR LOCAL PLUMBING GROUP PTY LTD        2016603   
1662                                 ZAMMIT, EDWARD      199613986   
1663                                   ZHANG, CHENG        2021181   
1664                            ZINGLEMANN, GREGORY      

## Checking the shape (rows & columns) of the final dataframe

In [66]:
df.shape

(1669, 5)

## Converting the dataframe into a `.csv` file

In [67]:
df.to_csv('ACT_Gasfitter.csv')